File: study0/solutions/week1/unique_binary_search_trees.py
# https://leetcode.com/problems/unique-binary-search-trees/description/

def numTreesRecursive(n, memoize):
    if n in memoize:
        return memoize[n]
    memoize_sum = 0
    for i in range(n):
        memoize_sum += numTreesRecursive(i, memoize) * numTreesRecursive(n - 1 - i, memoize)
    memoize[n] = memoize_sum
    return memoize_sum

class Solution:
    def numTrees(self, n: int) -> int:
        memoize = {0:1, 1:1}
        return numTreesRecursive(n, memoize)

File: study1/week1/unique-binary-search-trees.java
// https://leetcode.com/problems/unique-binary-search-trees/
// References: https://st-lab.tistory.com/300
class Solution {
    // Time Complexity: O(n^2)
    // Space Complexity: O(n^2)
    public int numTrees(int n) {
        int numOfTree = 0;

        if (n <= 1) {
            return 1;
        }

        // using recursion
        for (int i = 1; i <= n; ++i) {
            int leftSubTree = numTrees(i - 1);
            int rightSubTree = numTrees(n - i);
            numOfTree += leftSubTree * rightSubTree;
        }

        return numOfTree;
    }
}